File: src/WIP q3/q3.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

double isprime(double k)
{
  double largest_factor = 0;
  // Assuming 0 is non-prime
  if (k < 0 || k == 0 || k == 1 || (k % 2 == 0 && k != 2))
  {
    return 0;
  }
  else if (k == 2 || k == 3 || k == 5 || k == 7 || k == 11)
  {
    return 1;
  }
  for (int i = 3; i * i < k; i += 2)
  {
    if (fmod(k, i) == 0)
    {
      largest_factor = i;
    }
  }
  return largest_factor;
}
int main(void)
{
  double target = 600851475143;
  int primality = 0;
  while (primality == 0)
  {
    double is = isprime(target);
    if (is)
  }

  printf("%lf", is);
  return 0;
}

File: src/WIP q3/q3.py
from typing import Literal, Union
import math
import numpy as np
import numpy.typing as npt

target: float = 600851475143


def is_prime(number: float) -> Union[Literal[0, 1], float]:
    """
    Function to check whether a given number is prime
    ---
    """
    if number % 2 == 0 or number % 3 == -0 or number % 5 == 0:
        return 1

    upper_bound = math.floor(math.sqrt(number))
    # This if statement checks if it is even, if not, make it odd, since every
    # prime number except 2 is odd
    # we need only check every odd number later
    if upper_bound % 2 == 0:
        upper_bound -= 1
    for i in range(upper_bound, 2, -2):
        if upper_bound % i == 0:
            return i
    return 0


test = is_prime(target)
while test != 1 and test != 0:
    test = is_prime(test)

print(test)
